thirdparty/faiss/faiss/utils/structure-inl.h

```diff
@@ -0,0 +1,416 @@
+            const uint64_t& r13 = is_super ? b[13] : a13;
+            const uint64_t& r14 = is_super ? b[14] : a14;
+            const uint64_t& r15 = is_super ? b[15] : a15;
+            const uint64_t& r16 = is_super ? b[16] : a16;
+            const uint64_t& r17 = is_super ? b[17] : a17;
+            const uint64_t& r18 = is_super ? b[18] : a18;
+            const uint64_t& r19 = is_super ? b[19] : a19;
+            const uint64_t& r20 = is_super ? b[20] : a20;
+            const uint64_t& r21 = is_super ? b[21] : a21;
+            const uint64_t& r22 = is_super ? b[22] : a22;
+            const uint64_t& r23 = is_super ? b[23] : a23;
+            const uint64_t& r24 = is_super ? b[24] : a24;
+            const uint64_t& r25 = is_super ? b[25] : a25;
+            const uint64_t& r26 = is_super ? b[26] : a26;
+            const uint64_t& r27 = is_super ? b[27] : a27;
+            const uint64_t& r28 = is_super ? b[28] : a28;
+            const uint64_t& r29 = is_super ? b[29] : a29;
+            const uint64_t& r30 = is_super ? b[30] : a30;
+            const uint64_t& r31 = is_super ? b[31] : a31;
+            return (a0 & b[0]) == r0 && (a1 & b[1]) == r1 &&
+                   (a2 & b[2]) == r2 && (a3 & b[3]) == r3 &&
+                   (a4 & b[4]) == r4 && (a5 & b[5]) == r5 &&
+                   (a6 & b[6]) == r6 && (a7 & b[7]) == r7 &&
+                   (a8 & b[8]) == r8 && (a9 & b[9]) == r9 &&
+                   (a10 & b[10]) == r10 && (a11 & b[11]) == r11 &&
+                   (a12 & b[12]) == r12 && (a13 & b[13]) == r13 &&
+                   (a14 & b[14]) == r14 && (a15 & b[15]) == r15 &&
+                   (a16 & b[16]) == r16 && (a17 & b[17]) == r17 &&
+                   (a18 & b[18]) == r18 && (a19 & b[19]) == r19 &&
+                   (a20 & b[20]) == r20 && (a21 & b[21]) == r21 &&
+                   (a22 & b[22]) == r22 && (a23 & b[23]) == r23 &&
+                   (a24 & b[24]) == r24 && (a25 & b[25]) == r25 &&
+                   (a26 & b[26]) == r26 && (a27 & b[27]) == r27 &&
+                   (a28 & b[28]) == r28 && (a29 & b[29]) == r29 &&
+                   (a30 & b[30]) == r30 && (a31 & b[31]) == r31;
+        }
+    };
+
+    template<bool is_super>
+    struct StructureComputer512 {
+        uint64_t a0,a1,a2,a3,a4,a5,a6,a7,
+            a8,a9,a10,a11,a12,a13,a14,a15,
+            a16,a17,a18,a19,a20,a21,a22,a23,
+            a24,a25,a26,a27,a28,a29,a30,a31,
+            a32,a33,a34,a35,a36,a37,a38,a39,
+            a40,a41,a42,a43,a44,a45,a46,a47,
+            a48,a49,a50,a51,a52,a53,a54,a55,
+            a56,a57,a58,a59,a60,a61,a62,a63;
+
+        StructureComputer512() {}
+
+        StructureComputer512(const uint8_t *a8, int code_size) {
+            set(a8, code_size);
+        }
+
+        void set(const uint8_t *au8, int code_size) {
+            assert(code_size == 512);
+            const uint64_t *a = (uint64_t *)au8;
+            a0 = a[0]; a1 = a[1]; a2 = a[2]; a3 = a[3];
```

Comment:
why not use an array here?

Reply:
since #1647, the code is like this authored by shengjun.
I think it's because accessing a0 is faster than accessing "a[0]"